File: Plugins/TurnBasedUI/Source/TurnBasedUI/Private/Interop/ActionRegistrationExporter.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "Interop/ActionRegistrationExporter.h"

#include "CommonUserWidget.h"

void UActionRegistrationExporter::GetActionBindings(const UCommonUserWidget* Widget, const FUIActionBindingHandle*& Bindings,
                                                    int32& NumBindings)
{
    auto &AllBindings = Widget->GetActionBindings();
    Bindings = AllBindings.GetData();
    NumBindings = AllBindings.Num();  
}

void UActionRegistrationExporter::RegisterActionBinding(UCommonUserWidget* Widget, const FBindUIActionArgs& Args,
    FUIActionBindingHandle* Handle)
{
    *Handle = Widget->RegisterUIActionBinding(Args);   
}

void UActionRegistrationExporter::AddActionBinding(UCommonUserWidget* Widget, FUIActionBindingHandle* Handle)
{
    Widget->AddActionBinding(MoveTemp(*Handle));
}

void UActionRegistrationExporter::RemoveActionBinding(UCommonUserWidget* Widget, FUIActionBindingHandle* Handle)
{
    Widget->RemoveActionBinding(MoveTemp(*Handle));
}

void UActionRegistrationExporter::DestructHandle(FUIActionBindingHandle* Handle)
{
    std::destroy_at(Handle); 
}

File: Plugins/TurnBasedUI/Script/TurnBased.UI/Interop/ActionRegistrationExporter.cs
using TurnBased.UI.Actions;
using UnrealSharp.Binds;

namespace TurnBased.UI.Interop;

[NativeCallbacks]
public static unsafe partial class ActionRegistrationExporter
{
    private static readonly delegate* unmanaged<IntPtr, out IntPtr, out int, void> GetActionBindings;
    private static readonly delegate* unmanaged<IntPtr, ref FBindUIActionArgs, IntPtr, void> RegisterActionBinding;
    private static readonly delegate* unmanaged<IntPtr, IntPtr, void> AddActionBinding;
    private static readonly delegate* unmanaged<IntPtr, IntPtr, void> RemoveActionBinding;
    private static readonly delegate* unmanaged<IntPtr, void> DestructHandle;
}